feeluown/player/mpvplayer.py

```diff
@@ -152,20 +158,62 @@ def set_play_range(self, start=None, end=None):
             self.seeked.emit(start)
         _mpv_set_option_string(self._mpv.handle, b'end', bytes(end_str, 'utf-8'))
 
+    def fade(self, fade_in: bool, max_volume=None):
+        # k: factor between 0 and 1, to represent tick/fade_time
+        def fade_curve(k: float, fade_in: bool) -> float:
+            if fade_in:
+                return (1-math.cos(k*math.pi)) / 2
+            else:
+                return (1+math.cos(k*math.pi)) / 2
+
+        def set_volume(max_volume: int, fade_in: bool):
+            # https://bugs.python.org/issue31539#msg302699
+            if os.name == "nt":
+                freq = 25
+                interval = 0.02
+            else:
+                freq = 50
+                interval = 0.01
+
+            for _tick in range(freq):
+                new_volume = math.ceil(
+                    fade_curve(_tick/freq, fade_in=fade_in)*max_volume
+                )
+                self.volume = new_volume
+                time.sleep(interval)
```

Comment:
time.sleep blocks the current thread, so it may cause the UI hanging.
Consider using asyncio.sleep instead